File: tests/fixtures/go_repo/internal/types/user_dto.go
// Package types — request/response shapes for /users endpoints.
package types

import "time"

// CreateUserRequest is the JSON body for POST /users.
type CreateUserRequest struct {
	Email    string `json:"email" validate:"required,email"`
	Username string `json:"username" validate:"required,min=3,max=32"`
	Password string `json:"password" validate:"required,min=8"`
}

// UpdateUserRequest is the JSON body for PATCH /users/{id}. All fields are
// optional; nil pointers mean "leave unchanged".
type UpdateUserRequest struct {
	Email    *string `json:"email,omitempty" validate:"omitempty,email"`
	Username *string `json:"username,omitempty" validate:"omitempty,min=3,max=32"`
	Password *string `json:"password,omitempty" validate:"omitempty,min=8"`
}

// UserResponse is the public, serialised shape of a User. Never contains
// the password hash.
type UserResponse struct {
	ID        string    `json:"id"`
	Email     string    `json:"email"`
	Username  string    `json:"username"`
	CreatedAt time.Time `json:"created_at"`
}

// UserListResponse wraps a page of users with pagination metadata.
type UserListResponse struct {
	Items      []UserResponse `json:"items"`
	TotalCount int            `json:"total_count"`
	Page       int            `json:"page"`
	PageSize   int            `json:"page_size"`
}

File: tests/fixtures/rust_repo/src/dto/user_dto.rs
//! Request / response types for `/users` endpoints.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;
use validator::Validate;

/// JSON body for `POST /users`. All fields are required.
#[derive(Debug, Clone, Deserialize, Validate)]
pub struct CreateUserRequest {
    #[validate(email)]
    pub email: String,
    #[validate(length(min = 3, max = 32))]
    pub username: String,
    #[validate(length(min = 8))]
    pub password: String,
}

/// JSON body for `PATCH /users/{id}`. Every field is optional; `None`
/// means "leave unchanged".
#[derive(Debug, Clone, Deserialize, Validate)]
pub struct UpdateUserRequest {
    #[validate(email)]
    pub email: Option<String>,
    #[validate(length(min = 3, max = 32))]
    pub username: Option<String>,
    #[validate(length(min = 8))]
    pub password: Option<String>,
}

/// Public-facing user shape. Deliberately excludes `password_hash` so it can
/// never leak through this type.
#[derive(Debug, Clone, Serialize)]
pub struct UserResponse {
    pub id: Uuid,
    pub email: String,
    pub username: String,
    pub created_at: DateTime<Utc>,
}

/// Paginated wrapper used by `GET /users`.
#[derive(Debug, Clone, Serialize)]
pub struct UserListResponse {
    pub items: Vec<UserResponse>,
    pub total_count: i64,
    pub page: u32,
    pub page_size: u32,
}

impl From<crate::models::User> for UserResponse {
    fn from(u: crate::models::User) -> Self {
        UserResponse {
            id: u.id,
            email: u.email,
            username: u.username,
            created_at: u.created_at,
        }
    }
}
